### PR TITLE
CLOUDP-178750: Add deletion protection to deployments 

### PR DESCRIPTION
**Pull request by josvazg**
### All Submissions:

* [X] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

File: pkg/controller/atlasdeployment/atlasdeployment_controller.go [hunk 1]
```diff
@@ -52,13 +52,23 @@ import (
...
+	Scheme                      *runtime.Scheme
+	AtlasDomain                 string
+	GlobalAPISecret             client.ObjectKey
+	GlobalPredicates            []predicate.Predicate
+	EventRecorder               record.EventRecorder
+	ObjectDeletionProtection    bool
+	SubObjectDeletionProtection bool
+}
+
+type reconciliation struct {
```
Comment (helderjs) [inline pkg/controller/atlasdeployment/atlasdeployment_controller.go:66]: I found it hard to follow the reconciliation logic with this new struct. It doesn't make sense to add a new struct that will carry `AtlasDeploymentReconciler` and its dependencies inside of their own "methods".

**Review by helderjs — CHANGES_REQUESTED** [review]

File: pkg/controller/atlasdeployment/atlasdeployment_controller.go (hunk 1 above)
Comment (josvazg) [reply pkg/controller/atlasdeployment/atlasdeployment_controller.go:66]: I can refactor and nor make it a method, but the issue is that some pieces of code require a lot of context, that would be way too many parameters to pass ungrouped.

Comment (helderjs) [reply pkg/controller/atlasdeployment/atlasdeployment_controller.go:66]: You just moved the standalone functions to be pointer receivers of your new struct, you could also make them receivers of the reconciler itself. I also think tons of parameters in functions are a bad thing but you could group the parameters in a struct per function instead of a centralized struct that carries the parent and its dependencies.

Comment (josvazg) [reply pkg/controller/atlasdeployment/atlasdeployment_controller.go:66]: Ok, I will make that struct a deletion context struct only. All methods will return to the reconciler struct. Then I will just pass the deletion context struct as a first parameter of the method where appropriate, probably only in the new method.

Comment (josvazg) [reply pkg/controller/atlasdeployment/atlasdeployment_controller.go]: Done

**Comment by josvazg** [comment]
Ownership check is incomplete, will follow up on it soon.

**Comment by josvazg** [comment]
Ownership checks are now included, I wonder if we need any custom data transformation in the matching functions, like data format normalization.

**Comment by josvazg** [comment]
Just migrated unit tests to use Atlas service mocks instead of the http client mock.

File: pkg/controller/atlasdeployment/atlasdeployment_controller.go [hunk 2]
```diff
@@ -477,4 +562,108 @@ func (r *AtlasDeploymentReconciler) removeDeletionFinalizer(context context.Cont
...
+		isSame, err := deploymentMatchesSpec(log, typedAtlasCluster, deployment)
+		if err != nil {
+			return true, err
+		}
+
+		return !isSame, nil
+	}
+}
+
+func findTypedAtlasCluster(ctx context.Context, atlasClient mongodbatlas.Client, projectID, deploymentName string) (*atlasTypedCluster, error) {
```
[the remark this answers is not in the document]
Comment (josvazg) [reply pkg/controller/atlasdeployment/atlasdeployment_controller.go:602]: how do you suggest we fix it then?
at this point we only have a name, so we do not know what type of cluster we are looking for. In fact this call is to find either type by name and then succeed if any matched.

[the remark this answers is not in the document]
Comment (josvazg) [reply pkg/controller/atlasdeployment/atlasdeployment_controller.go:602]: also, how does atlas know to reply `CANNOT_USE_SERVERLESS_INSTANCE_IN_CLUSTER_API` if I only asked to GET by a name?

[the remark this answers is not in the document]
Comment (helderjs) [reply pkg/controller/atlasdeployment/atlasdeployment_controller.go:602]: It knows there's a cluster with the given name, but it's a serverless cluster, then it returns this error :man_shrugging: 
I'd also check for a 400 with this error code, not failing in this case, and let the code move to the next check.

[the remark this answers is not in the document]
Comment (josvazg) [reply pkg/controller/atlasdeployment/atlasdeployment_controller.go:602]: got it

File: pkg/controller/atlasdeployment/atlasdeployment_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+		log:         r.Log.With("atlasdeployment", "test-namespace"),
+		prevResult:  prevResult,
+		deployment:  deployment,
+		project:     project,
+	}
+}
+
+func testK8sClient() client.Client {
+	sch := runtime.NewScheme()
+	addSecretsListSchema(sch)
```
Comment (helderjs) [inline pkg/controller/atlasdeployment/atlasdeployment_controller_test.go:478]:
```suggestion
	s.AddKnownTypes(corev1.SchemeGroupVersion, &corev1.SecretList{})
```

File: pkg/controller/atlasdeployment/atlasdeployment_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+	return fake.NewClientBuilder().WithScheme(sch).Build()
+}
+
+func addSecretsListSchema(s *runtime.Scheme) {
+	s.AddKnownTypes(corev1.SchemeGroupVersion, &corev1.SecretList{})
+}
+
+func addDeploymentSchema(s *runtime.Scheme) {
+	s.AddKnownTypes(v1.GroupVersion, &v1.AtlasDeployment{})
+}
```
Comment (helderjs) [inline pkg/controller/atlasdeployment/atlasdeployment_controller_test.go:489]:
```suggestion
```

**Review by helderjs — CHANGES_REQUESTED** [review]

File: test/int/deployment_protected_test.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+		})
+
+		By("Stopping the operator", func() {
+			stopManager()
+			err := k8sClient.Delete(context.Background(), testNamespace)
+			Expect(err).ToNot(HaveOccurred())
+		})
+	})
+
+	It("removing from Kubernetes when protection is ON leaves the deployment in Atlas", func() {
```
[the remark this answers is not in the document]
Comment (josvazg) [reply test/int/deployment_protected_test.go:77]: So one case with serverless and one with advanced right?
ok, will do

File: pkg/controller/atlasdeployment/atlasdeployment_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+			Name:      fakeProject,
+			Namespace: ns,
+		},
+		Status: status.AtlasProjectStatus{
+			ID: fakeProjectID,
+		},
+	}
+}
+
+func testDeployment(project *v1.AtlasProject) *v1.AtlasDeployment {
```
[the remark this answers is not in the document]
Comment (josvazg) [reply pkg/controller/atlasdeployment/atlasdeployment_controller_test.go:523]: ok, will unwrap

[the remark this answers is not in the document]
Comment (josvazg) [reply pkg/controller/atlasdeployment/atlasdeployment_controller_test.go]: done

File: test/int/deployment_protected_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (josvazg) [reply test/int/deployment_protected_test.go]: done

**Review by helderjs — APPROVED** [review]

**Comment by josvazg** [comment]
The current CI issues with deployment-annotation-ns do not seem flakyness, there is an issue with ownership checking. It finds differences:
```diff
  	... // 13 identical fields
  	PitEnabled: &false,
  	StateName:  "IDLE",
  	ReplicationSpecs: []*mongodbatlas.AdvancedReplicationSpec{
  		&{
  			NumShards: 1,
  			ID:        "64d4091c69c4e9090c82f155",
  			ZoneName:  "Zone 1",
  			RegionConfigs: []*mongodbatlas.AdvancedRegionConfig{
  				&{
  					AnalyticsAutoScaling: nil,
- 					AnalyticsSpecs:       nil,
+ 					AnalyticsSpecs:       &mongodbatlas.Specs{InstanceSize: "M2", NodeCount: &0},
  					ElectableSpecs: &mongodbatlas.Specs{
  						DiskIOPS:      nil,
  						EbsVolumeType: "",
  						InstanceSize:  "M2",
- 						NodeCount:     nil,
+ 						NodeCount:     &3,
  					},
- 					ReadOnlySpecs:       nil,
+ 					ReadOnlySpecs:       &mongodbatlas.Specs{InstanceSize: "M2", NodeCount: &0},
  					AutoScaling:         nil,
  					BackingProviderName: "AWS",
  					... // 3 identical fields
  				},
  			},
  		},
  	},
  	CreateDate:   "2023-08-09T21:46:04Z",
  	RootCertType: "ISRGROOTX1",
  	... // 3 identical fields
```

I will check the test sequence in detail. In theory, this should not happen, as the operator should have taken control of the deployment, by annotating the previous config. Once that happens, it should not have to re-check for differences. Also, this deployment is created by the operator, so not sure how it can lose ownership like in this test. Will debug.

**Comment by josvazg** [comment]
Fix applied before I got the go.mod issue:
- The ownership check must only be affected by the protection configuration flag, not by the annotations to keep or not the object. That way protection off bypass all ownership checks, as expected.
- The test that was failing would fail with protection off. The Atlas deployment left there cannot be re-sync at redeployment time, as it will find differences and thus will fail to grab ownership of the object.

```diff
% git diff
diff --git a/pkg/controller/atlasdeployment/atlasdeployment_controller.go b/pkg/controller/atlasdeployment/atlasdeployment_controller.go
index 6276f9c..5aa1843 100644
--- a/pkg/controller/atlasdeployment/atlasdeployment_controller.go
+++ b/pkg/controller/atlasdeployment/atlasdeployment_controller.go
@@ -236,13 +236,13 @@ func (r *AtlasDeploymentReconciler) checkDeploymentIsManaged(
                dply.Spec.DeploymentSpec = nil
        }
 
-       isProtected := customresource.IsResourceProtected(dply, r.ObjectDeletionProtection)
        owner, err := customresource.IsOwner(
                dply,
-               isProtected,
+               r.ObjectDeletionProtection,
                customresource.IsResourceManagedByOperator,
                managedByAtlas(context, workflowCtx.Client, project.ID(), log),
        )
+
        if err != nil {
                result := workflow.Terminate(workflow.Internal, fmt.Sprintf("unable to resolve ownership for deletion protection: %s", err))
                workflowCtx.SetConditionFromResult(status.DatabaseUserReadyType, result)
diff --git a/test/e2e/annotations_test.go b/test/e2e/annotations_test.go
index 03b895b..84a71c9 100644
--- a/test/e2e/annotations_test.go
+++ b/test/e2e/annotations_test.go
@@ -31,6 +31,7 @@ var _ = Describe("Annotations base test.", Label("deployment-annotations-ns"), f
                        testData = test
                        mainCycle(test)
                },
+               // TODO: fix test for deletion protection on, as it would fail to re-take the cluster after deletion
                Entry("Simple configuration with keep resource policy annotation on deployment", Label("ns-crd"),
                        model.DataProvider(
                                "operator-ns-crd",
```

**Comment by josvazg** [comment]
`go.mod` issue fixed (missed a `go mod tidy`)

**Review by helderjs — APPROVED** [review]

**Review by igor-karpukhin — APPROVED** [review]
Looks good 👍

**Review by roothorp — APPROVED** [review]
Love all the testing :D
